src/box/iproto.cc

```diff
@@ -1540,14 +1736,92 @@ tx_process1(struct cmsg *m)
 	out = msg->connection->tx.p_obuf;
 	if (iproto_prepare_select(out, &svp) != 0)
 		goto error;
+
+	/*
+	 * Index operations can't begin a new multistatement
+	 * transaction.
+	 */
+	assert(in_txn() == txn);
+	fiber_set_txn(fiber(), NULL);
+
 	if (tuple && tuple_to_obuf(tuple, out))
 		goto error;
 	iproto_reply_select(out, &svp, msg->header.sync, ::schema_version,
 			    tuple != 0);
 	iproto_wpos_create(&msg->wpos, out);
 	return;
 error:
+	fiber_set_txn(fiber(), NULL);
+	tx_reply_error(msg);
+}
+
+static void
+tx_process_rollback_on_disconnect(struct cmsg *m)
+{
+	struct iproto_msg *msg = (struct iproto_msg *) m;
+	struct remote_txn *remote_txn = msg->txn;
+	assert(remote_txn->txn != NULL);
+	tx_prepare_transaction_for_request(remote_txn);
+	box_txn_rollback();
+}
+
+static void
+tx_process_remote_txn(struct cmsg *m)
+{
+	struct iproto_msg *msg = (struct iproto_msg *) m;
+	struct obuf *out = msg->connection->tx.p_obuf;
+	struct remote_txn *remote_txn = msg->txn;
+	enum iproto_type type = (enum iproto_type) msg->header.type;
+	int rc;
+
+	tx_fiber_init(msg->connection->session, msg->header.sync);
+	if (tx_check_schema(msg->header.schema_version))
+		goto error;
+
+	if (type == IPROTO_BEGIN) {
+		assert(remote_txn != NULL);
+		if (remote_txn->txn != NULL) {
+			diag_set(ClientError, ER_ACTIVE_TRANSACTION);
+			goto error;
+		}
+
+		rc = box_txn_begin();
+		if (rc != 0)
+			goto error;
+		struct txn *txn = in_txn();
+		assert(txn != NULL);
+		remote_txn->txn = txn;
+		trigger_clear(&txn->fiber_on_stop);
+		goto ok;
+	}
+	/*
+	 * The user can commit or rollback without errors, if
+	 * there is no transaction.
+	 */
+	if (remote_txn == NULL) {
+		goto ok;
+	}
+	txn = remote_txn->txn;
```

Comment:
here txn no name type name (expected unqualified-id before ‘=’ token), may be if it's used only for assert write
assert(remote_txn->txn != NULL)